### ISSUE TITLE
Best practice to add 'runtime' arguments to a Phery request?

### ISSUE THREAD

**Issue by holtkamp** [pocesar/phery#23]
I am currently looking at Phery to replace Xajax. 

Using Xajax we used something like: ```xajaxJavascriptFunctionName("stringArgument", 2, $("#elementId").data());```

Notice the last parameter, this would assemble an array of all HTML5 data attributes of a specific HTML element at runtime and hand them over to the PHP function invoked at the backend. 

How would this be possible with Phery, after some experiments I am wondering what would be considered a best practice?

I tried (no success):
```html
<button data-phery-remote="pheryFunctionName" data-phery-target="/ajax/process" data-phery-args=$("#elementId").data();>Submit request with runtime arguments</button>
```

Then tried (works, but obviously two requests are fired to `/ajax/process`)

```html
<button data-phery-remote="pheryFunctionName" data-phery-target="/ajax/process" data-phery-args="staticArgument" onclick="$(this).phery().remote($('#elementId').data());">Submit request with runtime arguments using onclick</button>
```

Then tried to append the run-time arguments (to potentially used static arguments) before firing the request:

```html
<button data-phery-remote="pheryFunctionName" data-phery-target="/ajax/process" data-phery-args="staticArgument" onclick="$(this).phery().append_args($('#elementId').data());">Submit request with appended runtime arguments using onclick()</button>
```

The last approach seems to work, but is this the advised way to realise sending over 'runtime' values as arguments to the PHP backend? It is not 'unobtrusive' anymore. Maybe I am looking for some kind of unobtrusive `data-phery-args-appender="$('#elementId').data()"` or  `data-phery-args-setter="$('#elementId').data()", but did not find it...

**Comment by pocesar** [comment]
there is no interpolation of parameters set in the HTML (like angular does), would require listening on the `phery:params` event on the element or in the document, then filtering per element.

```js
phery.on('params', function(event, obj){
// triggers for all elements
$.extend(obj, $('.element').data());
});
// or
$('.element').on('phery:params', function(event, obj){
// binds to only the selected elements
});
// or 'live' style
$(document).on('phery:params', '.element', function(event, obj){
});
```
there's also `data-phery-related` that takes a jQuery selector string, and merge multiple elements in the call

**Comment by holtkamp** [comment]
Ok thanks for the response. 

So in summary my ```onclick="$(this).phery().append_args($('#elementId').data());"``` approach would be appropriate? It does not need any event listener, so the impact on existing (Xajax-targeted) code would be minimized.

Would it be an option to introduce the suggested new functionality to append/set arguments upon dispatching the Phery request like:
- ```data-phery-args-appender="$('#elementId').data()"```
- ```data-phery-args-setter="$('#elementId').data()"```

Can try to come up with a PR, my Javascript is not so good however...

**Comment by pocesar** [comment]
yeah that solution sounds reasonable. the only problem with setting javascript inside the HTML is that it can be tampered quite easily, and would require `eval`. I guess it could ASSUME that it want data from an element, so acting like data-phery-related, passing a selector, would be enough.
would become:

```html
<button data-phery-args-append="#elementId" data-phery-related="#elementId"></button>
<button data-phery-args-append="this"></button>
``` 
setting to `this` would assume we want to append the current jQuery data on the element with any other data that phery needs to send

**Comment by holtkamp** [comment]
Ok, just a small note (downside?) about this approach, so this works:

```javascript
<button onclick="$(this).phery().append_args('parameter1');" data-phery-remote="remoteFunction" data-param="value" >Invoke Phery</button>
<button onclick="$(this).phery().append_args($('#element1').data());" data-phery-remote="remoteFunction" data-param="value" >Invoke Phery</button>
<button onclick="$(this).phery().append_args(JSON.stringify($(this).data()));" data-phery-remote="remoteFunction" data-param="value" >Invoke Phery</button>
```

But this approach results in a 'too much recursion' error, because of the infinite loop caused by the onclick event propagation:

```javascript
<button onclick="$(this).phery().append_args($(this).data());" data-phery-remote="remoteFunction" data-param="value" >Invoke Phery</button>
```

By stopping the event propagation, the Phery remote function is never invoked

```javascript
<button onclick="$(this).phery().append_args($(this).data());event.stopPropagation" data-phery-remote="remoteFunction" data-param="value" >Invoke Phery</button>.
```

So when sending over the data object as a JSON string, this would not be a problem, but when trying to send a native Javascript object, recursion occurs. Any idea how to tackle this, or is this just a consequence of merging arguments onclick...?

**Comment by pocesar** [comment]
This would have to be implemented as a new feature, because the element `onclick` has precedence over event listeners.
